### DISCUSSION TITLE
ScriptHookVDotNet acting weird and cuasing random game crashes

### DISCUSSION THREAD

**Discussion by MK77P** [scripthookvdotnet/scripthookvdotnet#1316] [Q&A]
I'm experiencing something really weird with this, after removing any mod that disables the despawn of DLC vehicles the DLC vehicles kept spawning, so I had to investigate which other mod is also disabling the despawn of DLC vehicles, it turned out that ScriptHookVDotNet would somehow disable the despawn of DLC vehicles if there was a script mod in the "scripts" folder (even if it was just one script mod) and if I removed that one last script from the "scripts" folder it would no longer disable the despawn of DLC vehicles.

Plus this is causing random crashes where the game freezes and the audio is in a loop then the game exists without any error message, not even in the Event Viewer.

**Comment by kagikn** [comment]
Could you tell me which SHVDN version (such as v3.6.0 or some commit hash), game version, and what scripts you were using when you experienced your "crash" where the game freezes (well it's not actually a crash for some exception such as access violation)? I'm assuming you are using SHVDN v3.6.0 because we started to initialize our memory stuff before scripts access some of them so a script won't be aborted for unintended timeout since 5f514ef, where the change is not included in v3.6.0.

SHVDN disables the despawn of DLC vehicles **on purpose** since you won't see any differences about DLC vehicles if you don't spawn any of them. Even we updated the countermeasure to support v1.0.2802.0 or later in v3.6.0. You may see weird behavior about whether SHVDN disables the despawn of DLC vehicles in v3.6.0 or earlier because it's part of our memory stuff initialization and it was invoked only whan some script access some of our memory stuff until v3.6.0.

**Reply by MK77P** [reply]
I'm using latest version of the game and SHVDN 3.6.0, these are the scripts that I use:

![Untitled](https://github.com/scripthookvdotnet/scripthookvdotnet/assets/145700160/9c001723-81fe-4c74-b149-b2e1f316cf5e)

While trying to locate the cause of the issue I would remove after each crash a few of the scripts, the result was the game crashes as long as there are scripts in the "scripts" folder as I kept swapping the scripts but still without any luck, I've completely removed SHVDN and not one single problem. 

I have DLC vehicles in popgroups, as I explained above, when there's any mod inside the "scripts" folder SHVDN disables the despawn of DLC vehicles, if I were to remove all mods from the "scripts" folder it stops disabling the despawn of DLC vehicles.

**Reply by kagikn** [reply]
You could try having one script loaded to find out the cause of the crash, one by one... So you could find the one(s) the crash or freeze causes in your game faster. I'm not a guy who is willing to decompile or disassemble all of the scripts assemblys of the scripts you are using (like 25 script assemblies I guess?) to diagnose your issue, either. It would be the best if you could provide sufficient info about how to reproduce your issue.
Since you seem to put scripts in subfolders, I must tell you an inportant note about how SHVDN scripts load their resources. **Lots** of SHVDN scripts naively assume that you have put them in the scripts folder, not expecting to put in a subfolder or another folder for the root script folder (SHVDN actually has the "ScriptsLocation" option for the ini). Follow what the installation instructions of scripts say if you are not sure about how scripts load their resources such as ini files, then you can blame the authors for not telling the correct info if the instructions says some false info.

The game version (such as v1.0.2944.0) you're using is also an important info to diagnose the issue, since SHVDN uses on a lot of memory stuff that game updates can break (if some scripts use some of our memory stuff). It would be appreciated if you could report. You can view one by viewing the property of the exe or seeing ScriptHookV.log. Some scripts may get not compatible for newer game versions due to their own memory stuff, though, you'll need to consult the authors if you suspect such case.

While you are trying to find the cause of the crash/freeze, the script contained in the archive below may be useful to prevent DLC vehicles from despawning (in v3.6.0 or earlier)... Shows an notification for kind of debugging purpose, but keep it placed until you find the cause of the crash/freeze. If you suffer some game crash even if you have only this script installed, I'd suggest you should use an nightly build at GitHub Actions tab. Search for a green check mark and download an artifact, then extract the asi, dlls, and the ini into the GTA V folder (you can ignore the xml files, they are for developers).
[ShvdnMemoryInit.3.zip](https://github.com/scripthookvdotnet/scripthookvdotnet/files/12695788/ShvdnMemoryInit.3.zip)

**Reply by kagikn** [reply]
I think I saw another message that seems to be deleted now. Since the deleted message says like you encountered the random crash with SHVDN installed but no SHVDN scripts, I'll ask another question. When you encountered a random crash with SHVDN installed but any SHVDN scripts, what error code `launcher.log` in `[Documents folder]\Rockstar Games\Launcher` says at the line `[gamelaunch] Game exited with code`? If your crash is not due to long freeze and using v3.6.0 or earlier ones, it can be 0xc00000fd, which may be really responsible to SHVDN (dde6861 mentions about this). Try an nightly build such as [this latest one (https://github.com/scripthookvdotnet/scripthookvdotnet/actions/runs/6263270547), especially if your game can crash with the exit code 0xc00000fd. It may resolve your issue if SHVDN v3.6.0 is causing it.
But if you have other mods not limited to script stuff installed, make sure some of them don't have major issues. Your vehicle mods or configs may be suspicious.

**Reply by MK77P** [reply]
Yeah to be honest I just gave up and deleted the message so I don't waste your time, I've been playing the game just fine for over two days without any crashes whatsoever when simply not using SHVDN along its scripts, but today thought about adding it back and trying to see if the crash is actually from SHVDN itself or from the scripts, so I installed them and the crashes came back directly, then thought about only keeping SHVDN without installing any mods, the game kept crashing with SHVDN only, then removed SHVDN and it kept crashing, so I have no idea what is going on here, the game had zero crashes in two days straight where I kept playing for hours and hours without no problems at all but as soon as I installed SHVDN and the scripts all the crashes returned then removed them all and the crashes still didn't stop.

[ ERROR ] [10632] [crashdetection] Exit code 0xc0000005 indicates a fatal game exit (reason: STATUS_ACCESS_VIOLATION)

This thing is going to give me cancer with its random crashes, I'm giving up trying to find the cause because it seems impossible. 

Thanks for all the help, appreciate it.
